### PR TITLE
Use PDBs with `unhealthyPodEvictionPolicy: AlwaysAllow` for Gardener managed components

### LINKED ISSUE #8821: Use PDBs with `unhealthyPodEvictionPolicy`
**How to categorize this issue?**

/area robustness
/area high-availability
/kind enhancement

**What would you like to be added**:
Since Kubernetes 1.27 (`PDBUnhealthyPodEvictionPolicy` defaulted to `true`) PDBs offer a [Unhealthy Pod Eviction Policy](https://kubernetes.io/docs/tasks/run-application/configure-pdb/#unhealthy-pod-eviction-policy).
It should be considered using `unhealthyPodEvictionPolicy: AlwaysAllow` for (some) Gardener managed components.

**Why is this needed**:
The default behavior for PDBs (also before 1.27) considers running pods (`.status.phase="Running"`) as potential candidates violating against a PDB. This becomes a problem if a pod of a single replica deployment is in `Running` but not `Ready` state and has a matching PDB with `maxUnavailable: 1` as endorsed by Gardener ([ref](https://github.com/gardener/gardener/blob/master/docs/development/high-availability.md#convenient-application-of-these-rules)).

For example, we saw `kube-controller-manager` pods being not `Ready` and thus blocked node roll-outs on seeds since the defined PDB was violated.
The [Shoot HA Best Practices](https://github.com/gardener/gardener/blob/master/docs/usage/shoot_high_availability_best_practices.md) documentation as well advertises advantages using `AlwaysAllow` in case of a zone outage.

### PR DESCRIPTION
**Pull request by shafeeqes**
**How to categorize this PR?**

/area robustness high-availability
/kind enhancement

**What this PR does / why we need it**:
This PR adapts Gardener managed components to use `unhealthyPodEvictionPolicy: AlwaysAllow` in their PDBs.
For more details, see https://github.com/gardener/gardener/issues/8821

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/gardener/issues/8821

**Special notes for your reviewer**:
I have adapted the components to use `unhealthyPodEvictionPolicy: AlwaysAllow` if the kubernetes version is >= `v1.26`, even though the featuregate is beta from only `v1.27`. This way, we can set the feature gate for the Shoots of the managedseed and the feature can be active from 1.26 itself. And similarly for the garden controlplane components. The changed PDB API is present from k8s v1.26.

NB: Hiding whitespace can improve review for changes in tests.

**Release note**:

```feature operator
The components managed by gardener now use PDBs with `unhealthyPodEvictionPolicy: AlwaysAllow` for clusters with kubernetes version >= 1.26. (For v1.26 clusters (shoots and virtual-garden cluster), the featuregate `PDBUnhealthyPodEvictionPolicy ` needs to be turned on in the kube-apiserver. From v1.27 this is enabled by default.)
```

### CODE REVIEW

File: pkg/component/hvpa/hvpa_test.go [hunk 1]
```diff
@@ -383,9 +394,25 @@ var _ = Describe("HVPA", func() {
...
+
+		Context("Kubernetes versions < 1.26", func() {
+			It("should successfully deploy all resources", func() {
+				Expect(string(managedResourceSecret.Data["poddisruptionbudget__"+namespace+"__hvpa-controller.yaml"])).To(Equal(componenttest.Serialize(podDisruptionBudgetFor(false))))
+			})
+		})
+
+		Context("Kubernetes versions >= 1.26", func() {
+			BeforeEach(func() {
+				values.KubernetesVersion = semver.MustParse("1.27.0")
```
Comment (ScheererJ) [inline pkg/component/hvpa/hvpa_test.go]: Should we not rather use some `1.26` version here?

File: pkg/component/clusterautoscaler/cluster_autoscaler.go [hunk 2]
```diff
@@ -242,12 +247,17 @@ func (c *clusterAutoscaler) Deploy(ctx context.Context) error {
...
+	unhealthyPodEvictionPolicyAlwatysAllow := policyv1.AlwaysAllow
 	if _, err := controllerutils.GetAndCreateOrMergePatch(ctx, c.client, podDisruptionBudget, func() error {
 		podDisruptionBudget.Labels = getLabels()
 		podDisruptionBudget.Spec = policyv1.PodDisruptionBudgetSpec{
 			MaxUnavailable: &pdbMaxUnavailable,
 			Selector:       deployment.Spec.Selector,
 		}
+		if versionutils.ConstraintK8sGreaterEqual126.Check(c.runtimeVersion) {
+			podDisruptionBudget.Spec.UnhealthyPodEvictionPolicy = &unhealthyPodEvictionPolicyAlwatysAllow
+		}
```
Comment (ScheererJ) [inline pkg/component/clusterautoscaler/cluster_autoscaler.go]: I am somewhat undecided about this. It is more or less replicated now all over the place. Should we not add a small helper function instead which gets the pod disruption budget and the version and adapts it accordingly?

File: pkg/component/etcd/bootstrap.go [hunk 3]
```diff
@@ -334,11 +336,16 @@ func (b *bootstrapper) Deploy(ctx context.Context) error {
...
-			podDisruptionBudget,
 			vpa,
 		}
 	)
 
+	if versionutils.ConstraintK8sGreaterEqual126.Check(b.kubernetesVersion) {
+		podDisruptionBudget.Spec.UnhealthyPodEvictionPolicy = &unhealthyPodEvictionPolicyAlwatysAllow
+	}
+
+	resourcesToAdd = append(resourcesToAdd, podDisruptionBudget)
```
Comment (ScheererJ) [inline pkg/component/etcd/bootstrap.go]: Any reason why `podDisruptionBudget` is now handled specially, i.e. added separately to `resourcesToAdd`? After the `var` block only one attribute is changed, but not the object reference itself.

File: pkg/component/istio/istio_test.go [hunk 4]
```diff
@@ -226,7 +236,7 @@ var _ = Describe("istiod", func() {
 		networkLabels = map[string]string{"to-target": "allowed"}
 
 		c = fake.NewClientBuilder().WithScheme(kubernetes.SeedScheme).Build()
-		renderer = chartrenderer.NewWithServerVersion(&version.Info{GitVersion: "v1.22.4"})
+		renderer = chartrenderer.NewWithServerVersion(&version.Info{GitVersion: "v1.27.2"})
```
Comment (ScheererJ) [inline pkg/component/istio/istio_test.go]: Should we stay on the boundary, i.e. use `1.26` here?

File: pkg/component/kubecontrollermanager/kube_controller_manager_test.go [hunk 5]
```diff
@@ -73,7 +73,7 @@ var _ = Describe("KubeControllerManager", func() {
 		namespace                     = "shoot--foo--bar"
 		version                       = "1.27.3"
 		semverVersion, _              = semver.NewVersion(version)
-		runtimeKubernetesVersion      = semver.MustParse("1.25.0")
+		runtimeKubernetesVersion      = semver.MustParse("1.27.3")
```
Comment (ScheererJ) [inline pkg/component/kubecontrollermanager/kube_controller_manager_test.go]: Again: Should we use `1.26` instead?

File: pkg/component/blackboxexporter/blackboxexporter_test.go [hunk 6]
```diff
@@ -82,7 +82,6 @@ var _ = Describe("BlackboxExporter", func() {
 
 	Describe("#Deploy", func() {
 		var (
-			vpaEnabled    = false
```
Comment (ScheererJ) [inline pkg/component/blackboxexporter/blackboxexporter_test.go:85]: Why did you remove this from the test?

File: pkg/component/gardenerapiserver/gardener_apiserver_test.go [hunk 7]
```diff
@@ -177,50 +177,71 @@ var _ = Describe("GardenerAPIServer", func() {
...
-					TargetPort: intstr.FromInt32(8443),
-				}},
-			},
+			}
+
+			if k8sGreaterEqual127 {
+				svc.Annotations["service.kubernetes.io/topology-mode"] = "auto"
+			} else {
+				svc.Annotations["service.kubernetes.io/topology-aware-hints"] = "auto"
+			}
```
Comment (ScheererJ) [inline pkg/component/gardenerapiserver/gardener_apiserver_test.go:240]: What is the reasoning for this change? It does not relate to the pod disruption budget story, right?

File: pkg/component/gardenerapiserver/gardener_apiserver_test.go [hunk 8]
```diff
@@ -1486,6 +1513,34 @@ kubeConfigFile: /etc/kubernetes/admission-kubeconfigs/validatingadmissionwebhook
...
+					It("should successfully deploy all resources", func() {
+						Expect(string(managedResourceSecretRuntime.Data["verticalpodautoscaler__some-namespace__gardener-apiserver-vpa.yaml"])).To(Equal(componenttest.Serialize(vpa)))
+						Expect(string(managedResourceSecretRuntime.Data["poddisruptionbudget__some-namespace__gardener-apiserver.yaml"])).To(Equal(componenttest.Serialize(podDisruptionBudgetFor(false))))
+						Expect(string(managedResourceSecretRuntime.Data["service__some-namespace__gardener-apiserver.yaml"])).To(Equal(componenttest.Serialize(serviceRuntimeFor(false))))
+					})
+				})
+
+				Context("when kubernetes version is < 1.27", func() {
+					BeforeEach(func() {
+						values.RuntimeVersion = semver.MustParse("1.25.0")
```
Comment (ScheererJ) [inline pkg/component/gardenerapiserver/gardener_apiserver_test.go]: Should this not be `1.26`? Otherwise it looks very similar to the block above.

File: pkg/component/gardenercontrollermanager/gardener_controller_manager_test.go [hunk 9]
```diff
@@ -90,7 +91,10 @@ var _ = Describe("GardenerControllerManager", func() {
 
 		fakeClient = fakeclient.NewClientBuilder().WithScheme(operatorclient.RuntimeScheme).Build()
 		fakeSecretManager = fakesecretsmanager.New(fakeClient, namespace)
-		values = Values{}
+
+		values = Values{
+			RuntimeVersion: semver.MustParse("1.27.1"),
```
Comment (ScheererJ) [inline pkg/component/gardenercontrollermanager/gardener_controller_manager_test.go]: Should we use `1.26` here?

File: pkg/component/gardenerscheduler/gardener_scheduler_test.go [hunk 10]
```diff
@@ -88,7 +89,9 @@ var _ = Describe("GardenerScheduler", func() {
 
 		fakeClient = fakeclient.NewClientBuilder().WithScheme(operatorclient.RuntimeScheme).Build()
 		fakeSecretManager = fakesecretsmanager.New(fakeClient, namespace)
-		values = Values{}
+		values = Values{
+			RuntimeVersion: semver.MustParse("1.27.1"),
```
Comment (ScheererJ) [inline pkg/component/gardenerscheduler/gardener_scheduler_test.go]: Why do we set the version here and overwrite it in the `BeforeEach(...)` below?

File: pkg/component/gardenerscheduler/gardener_scheduler_test.go [hunk 11]
```diff
@@ -284,7 +299,8 @@ var _ = Describe("GardenerScheduler", func() {
 			BeforeEach(func() {
 				// test with typical values
 				values = Values{
-					LogLevel: "info",
+					LogLevel:       "info",
+					RuntimeVersion: semver.MustParse("1.27.1"),
```
Comment (ScheererJ) [inline pkg/component/gardenerscheduler/gardener_scheduler_test.go]: Should we set this to `1.26`?

**Review by ScheererJ — COMMENTED** [review]
Thanks a lot for addressing this topic. It is structured nicely around the components.

I am not sure if staying directly on the boundaries would be better in terms of version, i.e. check with `1.25` and `1.26` instead of sometimes switching to `1.27`.

**Comment by shafeeqes** [comment]
> I am not sure if staying directly on the boundaries would be better in terms of version, i.e. check with 1.25 and 1.26 instead of sometimes switching to 1.27.

I chose this way because of the discussion in the issue, see https://github.com/gardener/gardener/issues/8821#issuecomment-1827529566. Also, for Shoots or Gardens, we can enable the feature gate explicitly and this would work. Otherwise we would have to adapt based on the featuregates and so on which is also an option. IMO, Since the API is present in 1.26, there would not be harm in enabling this. I do agree that this could cause confusion if one checks the PDB and the field is present, but it doesn't yield the desired result since the feature gate is off. 
@timuthy, WDYT?

File: pkg/component/gardenerapiserver/gardener_apiserver_test.go (hunk 7 above)
Comment (shafeeqes) [reply pkg/component/gardenerapiserver/gardener_apiserver_test.go:240]: No, just the tests were failing because I used 1.27 and thought I'd adapt it as well.

File: pkg/component/etcd/bootstrap.go (hunk 3 above)
Comment (shafeeqes) [reply pkg/component/etcd/bootstrap.go]: Just saw that it was done this way for service and deployment. I could change it if you want.

File: pkg/component/blackboxexporter/blackboxexporter_test.go (hunk 6 above)
Comment (shafeeqes) [reply pkg/component/blackboxexporter/blackboxexporter_test.go:85]: It is not required, the length is tested inside the Contexts for `w/` and `w/o` vpa.

File: pkg/component/gardenerscheduler/gardener_scheduler_test.go (hunk 10 above)
Comment (shafeeqes) [reply pkg/component/gardenerscheduler/gardener_scheduler_test.go]: We only overwrite it for `< 1.26` case. For the other case, this value is used.

File: pkg/component/etcd/bootstrap.go (hunk 3 above)
Comment (ScheererJ) [reply pkg/component/etcd/bootstrap.go]: I was just surprised by the somewhat unnecessary change, but if this is how it is handled in this file. Let's stay in sync with the remaining code.

File: pkg/component/gardenerscheduler/gardener_scheduler_test.go (hunk 10 above)
Comment (ScheererJ) [reply pkg/component/gardenerscheduler/gardener_scheduler_test.go]: My point was that we set it here and overwrite it in the common `BeforeEach(...)` in line 303. Then we overwrite it again in the `< 1.26` case in line 404.
I was just wondering if the tests are fine without this change as it seems to be overwritten anyway.
(It might only be relevant to the unchanged contexts `secrets`, `#Destroy` and `waiting functions`.)

Comment (shafeeqes) [reply pkg/component/gardenerscheduler/gardener_scheduler_test.go]: Yes, only for `secrets`, it's needed.

Comment (shafeeqes) [reply pkg/component/gardenerscheduler/gardener_scheduler_test.go]: https://github.com/gardener/gardener/pull/8969/commits/06c792a3c8af4ecb6a0352cb60023075f7c9519a should do the trick.

**Review by ScheererJ — COMMENTED** [review]
Thanks for the changes.

/lgtm

**Review by timuthy — COMMENTED** [review]
/assign

File: pkg/utils/kubernetes/poddisruptionbudget.go [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+import (
+	"github.com/Masterminds/semver/v3"
+	policyv1 "k8s.io/api/policy/v1"
+
+	versionutils "github.com/gardener/gardener/pkg/utils/version"
+)
+
+// SetUnhealthyPodEvictionPolicy sets the UnhealthyPodEvictionPolicy field if the kubernetes version is >= 1.26.
+func SetUnhealthyPodEvictionPolicy(pdb *policyv1.PodDisruptionBudget, kubernetesVersion *semver.Version) {
+	var unhealthyPodEvictionPolicyAlwatysAllow = policyv1.AlwaysAllow
```
Comment (timuthy) [inline pkg/utils/kubernetes/poddisruptionbudget.go]:
```suggestion
	var unhealthyPodEvictionPolicyAlwaysAllow = policyv1.AlwaysAllow
```

File: pkg/utils/kubernetes/poddisruptionbudget.go [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+import (
+	"github.com/Masterminds/semver/v3"
+	policyv1 "k8s.io/api/policy/v1"
+
+	versionutils "github.com/gardener/gardener/pkg/utils/version"
+)
+
+// SetUnhealthyPodEvictionPolicy sets the UnhealthyPodEvictionPolicy field if the kubernetes version is >= 1.26.
+func SetUnhealthyPodEvictionPolicy(pdb *policyv1.PodDisruptionBudget, kubernetesVersion *semver.Version) {
```
Comment (timuthy) [inline pkg/utils/kubernetes/poddisruptionbudget.go]:
```suggestion
func SetAlwaysAllowEviction(pdb *policyv1.PodDisruptionBudget, kubernetesVersion *semver.Version) {
```

Nit: To make it a more evident one-liner.

**Review by timuthy — CHANGES_REQUESTED** [review]
Nice, thank you very much for the implementation changes 🙏 
Can you please also adjust the documentation https://github.com/gardener/gardener/blob/master/docs/development/high-availability.md#convenient-application-of-these-rules?

**Comment by shafeeqes** [comment]
@timuthy PTAL!

**Review by timuthy — COMMENTED** [review]
/retest-required
/lgtm
/approve
